### PR TITLE
segtree update

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

might clean up BIT code later

### CODE REVIEW

File: content/4_Gold/PURS.mdx [hunk 1]
```diff
@@ -160,241 +160,300 @@ public class segtree {
...
-	void init(int _n) { n = _n; seg.assign(2*n,ID); }
-	void pull(int p) { seg[p] = comb(seg[2*p],seg[2*p+1]); }
-	void upd(int p, T val) { // set val at position p
-		seg[p += n] = val; for (p /= 2; p; p /= 2) pull(p); }
-	T query(int l, int r) {	// sum on interval [l, r]
-		T ra = ID, rb = ID;
-		for (l += n, r += n+1; l < r; l /= 2, r /= 2) {
-			if (l&1) ra = comb(ra,seg[l++]);
-			if (r&1) rb = comb(seg[--r],rb);
+		T range_min(int start, int end) {
```
Comment (ChuangSheep) [inline content/4_Gold/PURS.mdx:347]: maybe add short doc for inclusive/exclusive border? like [start, end] somehow

File: content/4_Gold/PURS.mdx [hunk 2]
```diff
@@ -160,241 +160,300 @@ public class segtree {
...
+			assert(0 <= start && start < len && 0 < end && end <= len);
+			T sum = DEFAULT;
+			for (start += len, end += len; start < end; start >>= 1, end >>= 1) {
+				if ((start & 1) != 0) {
+					sum = comb(sum, segtree[start++]);
+				}
+				if ((end & 1) != 0) {
+					sum = comb(sum, segtree[--end]);
+				}
```
Comment (ChuangSheep) [inline content/4_Gold/PURS.mdx:356]: r u sure it's `!= 0` not `== 0`? I mean i don't know much about segT... The original cpp and java code were already different though: 
image
image

File: content/4_Gold/PURS.mdx [hunk 3]
```diff
@@ -160,241 +160,300 @@ public class segtree {
...
+		const T DEFAULT = 0;
+
+		vector<T> segtree;
+		int len;
+	public:
+		SumSegmentTree(int len) : len(len), segtree(len * 2, DEFAULT) { }
+
+		void set(int ind, T val) {
+			assert(0 <= ind && ind < len);
+			for (segtree[ind += len] = val; ind > 1; ind >>= 1) {
```
Comment (ChuangSheep) [inline content/4_Gold/PURS.mdx:342]: I personal find `ind /= 2` more clear, but that's up to u
and maybe pull one of the assignment out of the for parentheses, the order of operations might be confusing

File: content/4_Gold/PURS.mdx [hunk 4]
```diff
@@ -160,241 +160,300 @@ public class segtree {
...
-			tree[index] = val;
-			for (index /= 2; index >= 1; index /= 2) {
-				tree[index] = tree[2*index] + tree[2*index+1];
+		for (int i = 0; i < queryNum; i++) {
+			int type = io.nextInt();
+			if (type == 1) {
+				int ind = io.nextInt() - 1;
+				int val = io.nextInt();
+				segtree.set(ind, val);
+			} else {
```
Comment (ChuangSheep) [inline content/4_Gold/PURS.mdx:445]:
```suggestion
			} else if (type == 2) {
```

**Review by ChuangSheep — COMMENTED** [review]
bitwise operators are just too much magic to me to understand

File: content/4_Gold/PURS.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/4_Gold/PURS.mdx:347]: that was in the previous implementation

File: content/4_Gold/PURS.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply content/4_Gold/PURS.mdx:356]: trust me, this works (you can test it yourself)

File: content/4_Gold/PURS.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply content/4_Gold/PURS.mdx:342]: tru tru

File: content/4_Gold/PURS.mdx (hunk 2 above)
Comment (ChuangSheep) [reply content/4_Gold/PURS.mdx:356]: then alright : )

File: content/4_Gold/PURS.mdx (hunk 1 above)
Comment (ChuangSheep) [reply content/4_Gold/PURS.mdx:347]: true, then it's ok
(though I still prefer to have it here as the previous is far away

File: content/4_Gold/PURS.mdx [hunk 5]
```diff
@@ -160,241 +160,307 @@ public class segtree {
...
+#include <iostream>
+#include <cassert>
+#include <vector>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+/** A data structure that can answer point update & range minimum queries. */
```
Comment (ryanchou-dev) [inline content/4_Gold/PURS.mdx:175]:
```suggestion
// A data structure that can answer point update & range minimum queries.
```

File: content/4_Gold/PURS.mdx [hunk 6]
```diff
@@ -160,241 +160,307 @@ public class segtree {
...
+template<class T>
+class MinSegmentTree {
+	private:
+		/** The operation to use for combining two elements. (Must be associative) */
```
Comment (ryanchou-dev) [inline content/4_Gold/PURS.mdx:179]: why are all your comments like this D:

File: content/4_Gold/PURS.mdx [hunk 7]
```diff
@@ -160,241 +160,307 @@ public class segtree {
...
-	for(int i=1; i<=q; i++) {
-		int t, a, b; cin >> t >> a >> b;
-		if (t==1) st.upd(a,b);
-		else cout << st.query(a,b) <<"\n";
+
+	for (int q = 0; q < query_num; q++) {
+		int type;
+		std::cin >> type;
+		if (type == 1) {
+			int ind, val;
```
Comment (ryanchou-dev) [inline content/4_Gold/PURS.mdx:230]: sep lines

**Review by ryanchou-dev — APPROVED** [review]
image

File: content/4_Gold/PURS.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply content/4_Gold/PURS.mdx:175]: https://stackoverflow.com/questions/20869421/is-there-a-way-to-write-a-single-line-javadoc-comment-like-in-doxygen

File: content/4_Gold/PURS.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply content/4_Gold/PURS.mdx:230]: thought they were closely related enough, but alr then

Comment (ryanchou-dev) [reply content/4_Gold/PURS.mdx:230]: you separate them in the other langs, so might as well be consistent

**Comment by SansPapyrus683** [comment]
will merge after deployment

**Pull request merged**